libraries/AC_Avoidance/AP_OADatabase.cpp

```diff
@@ -0,0 +1,277 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "AP_OADatabase.h"
+
+#if !HAL_MINIMIZE_FEATURES
+
+#include <AP_AHRS/AP_AHRS.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Math/AP_Math.h>
+#include <stdio.h>
+
+extern const AP_HAL::HAL& hal;
+
+#ifndef AP_OADATABASE_TIMEOUT_SECONDS_DEFAULT
+    #define AP_OADATABASE_TIMEOUT_SECONDS_DEFAULT   10
+#endif
+
+#ifndef AP_OADATABASE_SIZE_DEFAULT
+    #define AP_OADATABASE_SIZE_DEFAULT          100
+#endif
+
+#ifndef AP_OADATABASE_QUEUE_SIZE_DEFAULT
+    #define AP_OADATABASE_QUEUE_SIZE_DEFAULT 80
+#endif
+
+
+const AP_Param::GroupInfo AP_OADatabase::var_info[] = {
+
+    // @Param: SIZE
+    // @DisplayName: OADatabase maximum number of points
+    // @Description: OADatabase maximum number of points. Set to 0 to disable the OA Database.
+    // @Range: 0 10000
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("SIZE", 1, AP_OADatabase, _database_size_param, AP_OADATABASE_SIZE_DEFAULT),
+
+    // @Param: EXPIRE
+    // @DisplayName: OADatabase item timeout
+    // @Description: OADatabase item timeout. The time an item will linger without any updates before it expires. Zero means never expires which is useful for a sent-once static environment.
+    // @Units: s
+    // @Range: 0 127
+    // @Increment: 1
+    // @User: Advanced
+    AP_GROUPINFO("EXPIRE", 2, AP_OADatabase, _database_expiry_seconds, AP_OADATABASE_TIMEOUT_SECONDS_DEFAULT),
+
+    // @Param: QUEUE_SIZE
+    // @DisplayName: OADatabase queue maximum number of points
+    // @Description: OADatabase queue maximum number of points.
+    // @Range: 1 200
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("QUEUE_SIZE", 3, AP_OADatabase, _queue_size_param, AP_OADATABASE_QUEUE_SIZE_DEFAULT),
+AP_GROUPEND
+};
+
+AP_OADatabase::AP_OADatabase()
+{
+    if (_singleton != nullptr) {
+        AP_HAL::panic("AP_OADatabase must be singleton");
+    }
+    _singleton = this;
+
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_OADatabase::init()
+{
+    init_database();
+    init_queue();
+
+    if (_queue.items == nullptr || _database.items == nullptr) {
+        gcs().send_text(MAV_SEVERITY_INFO, "DB init failed . Sizes queue:%d, db:%d", _queue.size, _database.size);
+        free(_queue.items);
```

Comment:
Mixed new/free again here. This should be delete, or the new should be a variation of malloc.

Reply:
I'd like a second opinion

Reply:
per the dev call, all calloc()'s were removed and everything now uses new and delete